tools/module_dependencies.py

```diff
@@ -0,0 +1,127 @@
+__all__ = ['modules_dependent_on', 'dependency_graph']
+
+
+import numpy as np
+import importlib
+import multiprocessing
+import functools
+import sys
+import os
+import glob
+import pytest
+
+
+package = 'skimage'
+included_mods = [
+    # Add private and nested modules here
+    '_shared',
+    'filters.rank',
+]
+excluded_mods = [
+    '__version__',
+]
+
+
+@functools.cache
+def _pkg_modules() -> list[str]:
+    """List all package submodules.
+
+    Returns
+    -------
+    submodules : list
+        Sorted list of package submodules.
+    """
+    pkg = importlib.import_module(package)
+
+    members = {f'{package}.{attr}' for attr in dir(pkg)}
+    included_mods_full = {f'{package}.{mod}' for mod in included_mods}
+    excluded_mods_full = {f'{package}.{mod}' for mod in excluded_mods}
+
+    submodules = sorted((members | included_mods_full) - excluded_mods_full)
+
+    return submodules
+
+
+def _import_dependencies(module: str) -> set[str]:
+    # Check that process is fresh (we've had trouble with it being forked!)
+    bad_modules = {mod for mod in sys.modules if f"{package}." in mod}
+    if bad_modules:
+        msg = f"Expected a fresh process, but {bad_modules} were already imported."
+        raise RuntimeError(msg)
+
+    # Import the module
+    mod = importlib.import_module(module)
+
+    # Import its test modules
+    mod_dir = os.path.dirname(mod.__file__)
+    test_files = glob.glob('**/test_*.py', root_dir=mod_dir, recursive=True)
+    test_modules = [f"{module}/{f}".replace("/", ".")[:-3] for f in test_files]
+    for mod in test_modules:
+        try: importlib.import_module(mod)
+        except pytest.skip.Exception: pass  # raised by `pytest.importorskip`
+
+    # Return the modules that `module` depends on
+    return set(_pkg_modules()) & set(sys.modules)
+
+
+def dependency_graph():
+    """Calculate the dependency graph of the current module.
+
+    Each row represents the dependencies of one subpackage.
+
+    """
+    mods = _pkg_modules()
+
+    n = len(mods)
+    A = np.zeros((n, n), dtype=bool)
+
+    multiprocessing.set_start_method('spawn')
+
+    with multiprocessing.Pool(maxtasksperchild=1) as p:
+        mod_deps = p.imap(_import_dependencies, mods)
+        for k, dependencies in enumerate(mod_deps):
+            for mod in dependencies:
+                A[k, mods.index(mod)] = True
+
+    return A
+
+
+def dependency_toml():
+    mods = _pkg_modules()
+    mods_arr = np.array(mods, dtype=object)
+
+    A = dependency_graph()
+
+    toml = []
+    for i, mod in enumerate(mods):
+        dependency_names = mods_arr[A[:, i]].tolist()
+
+        toml.append({"modules": {"path": mod, "depends_on": dependency_names}})
```

Comment:
I think the English interpretation is reversed here. 
- Each row `i` indicates all the modules that module `i` depends on.
- Each column `j` indicates all the modules that depend on module `j`.
- We are extracting columns here, so the list of modules we get all depend on module `j`.
- So if "path" is module `j`, it is _depended on by_ all these modules.

```suggestion
        toml.append({"modules": {"path": mod, "depended_on_by": dependency_names}})
```
In other words, if we change module `j`, we need to run tests for the modules given by `dependency_names`. Right?  Alternatively:

```suggestion
        toml.append({"modules": {"path": mod, "dependency_of": dependency_names}})
```

So perhaps a better name for `dependency_names` is `dependent_names`, but really... to avoid confusion between "dependancy" and "dependent", maybe we need other terms that are easier to tell apart.

Reply:
This wording came from the SciPy PR; it's what their tool puts out. I thought it would be good to be compatible, but happy to talk about what this should look like.

Reply:
This comment can be closed; it's the code that needs to be changed.